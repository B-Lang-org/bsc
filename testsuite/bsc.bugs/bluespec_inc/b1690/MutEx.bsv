import Vector :: *;

(* synthesize *)
module mkMutEx();

   Vector#(2,Reg#(Bool)) flag <- replicateM(mkReg(True));
   Reg#(Bool)            x    <- mkRegU();

   rule r1 if (flag[0]);
      $display("R1");
      x <= !x;
      flag[0] <= False;
   endrule

// Compiler does not see mutual exclusivity of r1 and r2 with this:
   Bit#(2) prefix = pack(readVReg(flag)) & 2'b01;

// Compiler does see mutual exclusivity of r1 and r2 with this:
//   Bit#(2) prefix = {1'b0,pack(flag[0])};

   rule r2 if (flag[1] && prefix == '0);
      $display("R2");
      x <= !x;
      flag[1] <= False;
   endrule

   rule done if (pack(readVReg(flag)) == '0);
      $finish(0);
   endrule

endmodule
