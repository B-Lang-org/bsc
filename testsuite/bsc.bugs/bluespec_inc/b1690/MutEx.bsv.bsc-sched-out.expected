checking package dependencies
compiling MutEx.bsv
code generation for mkMutEx starts
=== schedule:
parallel: [esposito: [RL_r1 -> [], RL_r2 -> [], RL_done -> []]]
order: [RL_r1, RL_r2, RL_done]

-----

=== resources:
[(flag_0.read, [(flag_0.read, 1)]),
 (flag_0.write, [(flag_0.write 1'd0, 1)]),
 (flag_1.read, [(flag_1.read, 1)]),
 (flag_1.write, [(flag_1.write 1'd0, 1)]),
 (x.read, [(x.read, 1)]),
 (x.write, [(x.write NOT_x___d3, 1)])]

-----

=== vschedinfo:
SchedInfo [] [] [] []

-----

Schedule dump file created: mkMutEx.sched
=== Generated schedule for mkMutEx ===

Rule schedule
-------------
Rule: r1
Predicate: flag_0
Blocking rules: (none)
 
Rule: r2
Predicate: flag_1 && (! flag_0)
Blocking rules: (none)
 
Rule: done
Predicate: {flag_1, flag_0} == 2'd0
Blocking rules: (none)
 
Logical execution order: r1, r2, done

=======================================
Verilog file created: mkMutEx.v
All packages are up to date.
